README.md
# Geomodel

[![Build Status](https://secure.travis-ci.org/integrallis/geomodel.png?branch=master)](http://travis-ci.org/integrallis/geomodel) 
[![Gem Version](https://badge.fury.io/rb/geomodel.png)](http://badge.fury.io/rb/geomodel)
[![Dependency Status](https://gemnasium.com/integrallis/geomodel.png)](https://gemnasium.com/integrallis/geomodel) 
[![Code Climate](https://codeclimate.com/github/integrallis/geomodel.png)](https://codeclimate.com/github/integrallis/geomodel)
[![Coverage Status](https://coveralls.io/repos/integrallis/geomodel/badge.png?branch=master)](https://coveralls.io/r/integrallis/geomodel?branch=master)

Geomodel aims to provide a generalized solution for performing basic indexing 
and querying of geospatial data in non-relation environments. At the core, this 
solution utilizes geohash-like objects called geocells.

A geocell is a hexadecimal string that defines a two dimensional rectangular
region inside the [-90,90] x [-180,180] latitude/longitude space. A geocell's
'resolution' is its length. For most practical purposes, at high resolutions,
geocells can be treated as single points.

Much like geohashes (see http://en.wikipedia.org/wiki/Geohash), geocells are 
hierarchical, in that any prefix of a geocell is considered its ancestor, with
geocell[:-1] being geocell's immediate parent cell.

To calculate the rectangle of a given geocell string, first divide the
[-90,90] x [-180,180] latitude/longitude space evenly into a 4x4 grid like so:

<pre>
               +---+---+---+---+ (90, 180)
               | a | b | e | f |
               +---+---+---+---+
               | 8 | 9 | c | d |
               +---+---+---+---+
               | 2 | 3 | 6 | 7 |
               +---+---+---+---+
               | 0 | 1 | 4 | 5 |
    (-90,-180) +---+---+---+---+
</pre>

NOTE: The point (0, 0) is at the intersection of grid cells 3, 6, 9 and c. And,
for example, cell 7 should be the sub-rectangle from (-45, 90) to (0, 180).   

Calculate the sub-rectangle for the first character of the geocell string and
re-divide this sub-rectangle into another 4x4 grid. For example, if the geocell
string is '78a', we will re-divide the sub-rectangle like so:

<pre>
                 .                   .
                 .                   .
             . . +----+----+----+----+ (0, 180)
                 | 7a | 7b | 7e | 7f |
                 +----+----+----+----+
                 | 78 | 79 | 7c | 7d |
                 +----+----+----+----+
                 | 72 | 73 | 76 | 77 |
                 +----+----+----+----+
                 | 70 | 71 | 74 | 75 |
    . . (-45,90) +----+----+----+----+
                 .                   .
                 .                   .
</pre>

Continue to re-divide into sub-rectangles and 4x4 grids until the entire
geocell string has been exhausted. The final sub-rectangle is the rectangular
region for the geocell.

A geocell can be associated with a single geographic point and subsequently 
indexed and filtered by either conformance to a bounding box or by proximity 
(nearest-n) to a search center point.

# Approach

This Ruby implementation of GeoModel is based on the Python, Java and JavaScript implementations. 
It's implemented as class level methods contained within modules and a few datatype classes. So the 
'model' part isn't quite there and I don't really see a need for it. Since the library is meant to 
be use in Non-Relational/Non-ORM environmets, binding the functions/methods to a model does not make
much sense. 

The model part was mostly implemented in the other libraries to bind directly to Google App Engine.
The idea here is to make it backend/db independent and use callbacks to integrate with the backend.

# References

- http://code.google.com/p/javageomodel/
- http://code.google.com/p/geomodel/
- https://github.com/danieldkim/geomodel

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'geomodel'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install geomodel

## Usage

Currently, only single-point entities and two types of basic geospatial queries
on those entities are supported.

### Representing your Locations

You'll need a class to hold a geolocation. It assumes that an "entity" has a unique 
"id" (specific field can be configure), a latitude/longitude combination stored in 
a "location" field (a Geomodel::Types::Point) and a collection of "geocells".

```ruby
class Entity
  attr_accessor :id, :location, :geocells
  
  def to_s
    self.id
  end
end
```

An instance of one of these entities can be instantiated as shown next. Let's say we
wanted to create an entity for the Frank Lloyd Wright Iconic Desert Spire in Scottsdale, AZ
(http://livebetterinscottsdale.com/2012/02/things-to-see-in-scottsdale-az-the-frank-lloyd-wright-spire/):

```ruby
flw_spire = Entity.new
flw_spire.id = 'Flatiron'
flw_spire.location = Geomodel::Types::Point.new(33.633406, -111.916803)
flw_spire.geocells = Geomodel::GeoCell.generate_geocells(flw_spire.location)

puts flw_spire.geocells
# 8
# 8d
# 8da
# 8daa
# 8daab
# 8daab6
# 8daab66
# 8daab666
# 8daab6668
# 8daab66684
# 8daab66684e
# 8daab66684e4
# 8daab66684e4d
```

### Bounding Box Queries

```ruby
# compute a geocell for the location using a resolution of 14
cell = Geomodel::GeoCell.compute(flw_spire.location, 14)

# create a bounding box for the cell
bounding_box = Geomodel::GeoCell.compute_box(cell)

# get a list of geocells for the given bounding box
geocells = Geomodel.geocells_for_bounding_box(bounding_box)

# use the bounding box geocells to do a key lookup in your database assuming that there is
# a location_geocell 'column' and you can do an IN query like:
result_set = my_db.query('SELECT * WHERE location_geocells IN (?)', query_geocells)

# the results then can be filtered by whether they fall inside the bounding box using:
matches = Geomodel.filter_result_set_by_bounding_box(bounding_box, result_set)
```

### Proximity (nearest-n) Queries

Find nearby locations given a location (lat & lon) and a radius in meters:

```ruby

# a list of places (instance of Entity or object that responds to :id, :location, :geocells)
places = [place1, place2, place3, ...]

# a function that can query your database. It takes as a parameter an array of geocells (strings)
# that are used to filter the query (below is an in-memory implementation using the 'places' array
# as our datasource)
query_runner = lambda do |geocells|
  result = places.reject do |o| 
    (o.geocells & geocells).length < 0
  end 

  result
end

# query for a maximum of 20 results, 15 miles (~24140 meters) from the Frank Lloyd Wright Spire
# results are tuples (2 element arrays) with the matching entity and its distance from the location
results = Geomodel.proximity_fetch(flw_spire.location, query_runner, 20, 24140)

# extract the matching places
places = results.map(&:first)

# extract the distances
distances = results.map(&:last)

```

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## License

MIT License


CHANGELOG.md
### 0.0.1 (January 2, 2014)
  * Initial release - Straight port from Python, Java and JS implementations

### 0.0.2 (January 5, 2014)
  * Upped the test coverage (which revealed some pretty blatant bugs)

### 0.0.3 (January 6, 2014)
	* Cleaned up more bugs in proximity search after testing with Cassandra demo app

spec/util_spec.rb
require 'spec_helper'

describe 'Geomodel::Util' do
  
  it 'can merge and sort one or many arrays into a target array' do
    list1 = [0, 1, 5, 6, 8, 9, 15]
    list2 = [0, 2, 3, 5, 8, 10, 11, 17]
    list3 = [1, 4, 6, 8, 10, 15, 16]
    list4 = [-1, 19]
    list5 = [20]
    list6 = []
    
    
    Geomodel::Util.merge_in_place(list1, [list2, list3, 
                                  list4, list5, list6], lambda { |x| x })

    expect(list1).to eq([-1, 0, 1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 15, 16, 17, 19, 20])
  end
end

spec/spec_helper.rb
require 'simplecov'
require 'coveralls'
Coveralls.wear!

SimpleCov.command_name 'Unit Tests'
SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  SimpleCov::Formatter::HTMLFormatter,
  Coveralls::SimpleCov::Formatter
]
SimpleCov.start do
  add_filter "/spec/"
end

require 'rubygems'
require 'pry'

$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))

require 'geomodel'
require 'rspec'
require 'rspec/autorun'

RSpec.configure do |config|

end

lib/geomodel.rb
require "geomodel/version"
require "geomodel/geomath"
require "geomodel/geotypes"
require "geomodel/geocell"
require "geomodel/util"
require 'set'

module Geomodel
  
  # The default cost function, used if none is provided              
  DEFAULT_COST_FUNCTION = lambda do |num_cells, resolution|
    num_cells > (Geomodel::GeoCell::GEOCELL_GRID_SIZE ** 2) ? 1e10000 : 0
  end
  
  # Retrieve the geocells to be used in a bounding box query
  # Something like geocells IN (...)
  # 
  # Args:
  #
  #   bbox: A geotypes.Box indicating the bounding box to filter entities by.
  #   cost_function: An optional function that accepts two arguments:
  #       * num_cells: the number of cells to search
  #       * resolution: the resolution of each cell to search
  #       and returns the 'cost' of querying against this number of cells
  #       at the given resolution.
  def self.geocells_for_bounding_box(bounding_box, cost_function = nil)
    cost_function = DEFAULT_COST_FUNCTION if cost_function.nil?
    Geomodel::GeoCell.best_bbox_search_cells(bounding_box, cost_function)
  end
  
  # Given a result set from your datastore (a query you filtered with
  # #geocells_for_bounding_box) it will filter the records that land outside
  # of the given bounding box (generally you'll use the same bounding box used
  # in #geocells_for_bounding_box)
  def self.filter_result_set_by_bounding_box(bounding_box, result_set)
    result_set.select do |row|
      row.latitude >= bounding_box.south &&
      row.latitude <= bounding_box.north &&
      row.longitude >= bounding_box.west &&
      row.longitude <= bounding_box.east
    end
  end
  
  #   center: A geotypes.Point or db.GeoPt indicating the center point around
  #       which to search for matching entities.
  #   max_results: An int indicating the maximum number of desired results.
  #       The default is 10, and the larger this number, the longer the fetch
  #       will take.
  #   max_distance: An optional number indicating the maximum distance to
  #       search, in meters.
  def self.proximity_fetch(center, query_runner, max_results = 10, max_distance = 0)
    results = []
  
    searched_cells = Set.new
    
    # The current search geocell containing the lat,lon.
    cur_containing_geocell = Geomodel::GeoCell.compute(center)
    
    # The currently-being-searched geocells.
    # NOTES:
    #     * Start with max possible.
    #     * Must always be of the same resolution.
    #     * Must always form a rectangular region.
    #     * One of these must be equal to the cur_containing_geocell.
    cur_geocells = [cur_containing_geocell]
  
    closest_possible_next_result_dist = 0
    
    # Assumes both a and b are lists of (entity, dist) tuples, *sorted by dist*.
    # NOTE: This is an in-place merge, and there are guaranteed
    # no duplicates in the resulting list.
    
    cmp_fn = lambda do |x, y| 
      (!x.empty? && !y.empty?) ? x[1] <=> y[1] : 0
    end
    
    dup_fn = lambda do |x|
      (x.nil? || x.empty?) ? nil : x[0].id 
    end # assuming the the element responds to #id
  
    sorted_edges = [[0,0]]
    sorted_edge_distances = [0]
    
    while !cur_geocells.empty?
      closest_possible_next_result_dist = sorted_edge_distances[0]
      
      break if max_distance && closest_possible_next_result_dist > max_distance
  
      cur_geocells_unique = cur_geocells - searched_cells.to_a
        
      # Run query on the next set of geocells.
      cur_resolution = cur_geocells[0].size
      
      # Update results and sort.
      new_results = query_runner.call(cur_geocells_unique)
      
      searched_cells.merge(cur_geocells)
      
      # Begin storing distance from the search result entity to the
      # search center along with the search result itself, in a tuple.
      new_results = new_results.map { |entity|  [entity, Geomodel::Math.distance(center, entity.location)] }
      new_results.sort! { |x, y| (!x.empty? && !y.empty?) ? x[1] <=> y[1] : 0 }
      new_results = new_results[0...max_results]
      
      # Merge new_results into results or the other way around, depending on
      # which is larger.
      if results.size > new_results.size
        Geomodel::Util.merge_in_place(results, [new_results], dup_fn, cmp_fn)
      else
        Geomodel::Util.merge_in_place(new_results, [results], dup_fn, cmp_fn)
        results = new_results
      end

      results = results[0...max_results]
      
      sorted_edges, sorted_edge_distances = Geomodel::Util.distance_sorted_edges(cur_geocells, center)
  
      if results.empty? || cur_geocells.size == 4
        # Either no results (in which case we optimize by not looking at adjacents, go straight to the parent) 
        # or we've searched 4 adjacent geocells, in which case we should now search the parents of those
        # geocells.
        cur_containing_geocell = cur_containing_geocell[0...-1]
        cur_geocells = cur_geocells.map { |cell| cell[0...-1] }   
        break if cur_geocells.empty? || !cur_geocells[0] # Done with search, we've searched everywhere.
      elsif cur_geocells.size == 1
        # Get adjacent in one direction.
        # TODO(romannurik): Watch for +/- 90 degree latitude edge case geocells.
        nearest_edge = sorted_edges[0]
        cur_geocells << Geomodel::GeoCell.adjacent(cur_geocells[0], nearest_edge)
      elsif cur_geocells.size == 2
        # Get adjacents in perpendicular direction.
        nearest_edge = Geomodel::Util.distance_sorted_edges([cur_containing_geocell], center)[0][0]
        if nearest_edge[0] == 0
          # Was vertical, perpendicular is horizontal.
          perpendicular_nearest_edge = sorted_edges.keep_if { |x| x[0] != 0 }.first
        else
          # Was horizontal, perpendicular is vertical.
          perpendicular_nearest_edge = sorted_edges.keep_if { |x| x[0] == 0 }.first
        end
        
        cur_geocells.concat(
          cur_geocells.map { |cell| Geomodel::GeoCell.adjacent(cell, perpendicular_nearest_edge) } 
        )
      end
      
      # We don't have enough items yet, keep searching.
      next if results.size < max_results
  
      # If the currently max_results'th closest item is closer than any
      # of the next test geocells, we're done searching.
      current_farthest_returnable_result_dist = Geomodel::Math.distance(center, results[max_results - 1][0].location)
      break if (closest_possible_next_result_dist >= current_farthest_returnable_result_dist)
    end
    
    results[0...max_results].keep_if { |result| max_distance == 0 || result.last < max_distance }  
  end
  
end

lib/geomodel/geomath.rb
require 'geocoder'

module Geomodel::Math
  
  RADIUS = 6378135
  
  # Calculates the great circle distance between two points (law of cosines).
  # 
  # Args:
  #   start_point: A geotypes.Point or db.GeoPt indicating the first point.
  #   end_point_: A geotypes.Point or db.GeoPt indicating the second point.
  # 
  # Returns:
  #   The 2D great-circle distance between the two given points, in meters.
  #
  def self.distance(start_point, end_point)
    start_point_lat = Geocoder::Calculations.to_radians(start_point.latitude)
    start_point_lon = Geocoder::Calculations.to_radians(start_point.longitude)
    end_point_lat = Geocoder::Calculations.to_radians(end_point.latitude) 
    end_point_lon = Geocoder::Calculations.to_radians(end_point.longitude)
    # work out the internal value for the spherical law of cosines and clamp
    # it between -1.0 and 1.0 to avoid rounding errors
    sloc = (Math.sin(start_point_lat) * Math.sin(end_point_lat) +
            Math.cos(start_point_lat) * Math.cos(end_point_lat) * Math.cos(end_point_lon - start_point_lon))
    sloc = [[sloc, 1.0].min, -1.0].max
    RADIUS * Math.acos(sloc)
  end
end

lib/geomodel/geocell.rb
module Geomodel
  
  # Defines the notion of 'geocells' and exposes methods to operate on them.
  # 
  # A geocell is a hexadecimal string that defines a two dimensional rectangular
  # region inside the [-90,90] x [-180,180] latitude/longitude space. A geocell's
  # 'resolution' is its length. For most practical purposes, at high resolutions,
  # geocells can be treated as single points.
  # 
  # Much like geohashes [see http://en.wikipedia.org/wiki/Geohash], geocells are
  # hierarchical, in that any prefix of a geocell is considered its ancestor, with
  # geocell[:-1] being geocell's immediate parent cell.
  # 
  # To calculate the rectangle of a given geocell string, first divide the
  # [-90,90] x [-180,180] latitude/longitude space evenly into a 4x4 grid like so:
  # 
  #              +---+---+---+---+ [90, 180]
  #              | a | b | e | f |
  #              +---+---+---+---+
  #              | 8 | 9 | c | d |
  #              +---+---+---+---+
  #              | 2 | 3 | 6 | 7 |
  #              +---+---+---+---+
  #              | 0 | 1 | 4 | 5 |
  #   [-90,-180] +---+---+---+---+
  # 
  # NOTE: The point [0, 0] is at the intersection of grid cells 3, 6, 9 and c. And,
  #       for example, cell 7 should be the sub-rectangle from
  #       [-45, 90] to [0, 180].
  # 
  # Calculate the sub-rectangle for the first character of the geocell string and
  # re-divide this sub-rectangle into another 4x4 grid. For example, if the geocell
  # string is '78a', we will re-divide the sub-rectangle like so:
  # 
  #                .                   .
  #                .                   .
  #            . . +----+----+----+----+ [0, 180]
  #                | 7a | 7b | 7e | 7f |
  #                +----+----+----+----+
  #                | 78 | 79 | 7c | 7d |
  #                +----+----+----+----+
  #                | 72 | 73 | 76 | 77 |
  #                +----+----+----+----+
  #                | 70 | 71 | 74 | 75 |
  #   . . [-45,90] +----+----+----+----+
  #                .                   .
  #                .                   .
  # 
  # Continue to re-divide into sub-rectangles and 4x4 grids until the entire
  # geocell string has been exhausted. The final sub-rectangle is the rectangular
  # region for the geocell.
  #
  module GeoCell
    # Geocell algorithm constants.
    GEOCELL_GRID_SIZE = 4
    GEOCELL_ALPHABET = '0123456789abcdef'

    # The maximum *practical* geocell resolution.
    MAX_GEOCELL_RESOLUTION = 13

    # The maximum number of geocells to consider for a bounding box search.
    MAX_FEASIBLE_BBOX_SEARCH_CELLS = 300

    # Direction enumerations.
    NORTHWEST = [-1, 1]
    NORTH = [0, 1]
    NORTHEAST = [1, 1]
    EAST = [1, 0]
    SOUTHEAST = [1, -1]
    SOUTH = [0, -1]
    SOUTHWEST = [-1, -1]
    WEST = [-1, 0]
    
    def self.generate_geocells(point)
      geocell_max = self.compute(point, MAX_GEOCELL_RESOLUTION)
      
      (1..MAX_GEOCELL_RESOLUTION).map do |resolution|
        self.compute(point, resolution)
      end
    end
    
    # Returns an efficient set of geocells to search in a bounding box query.
    # 
    # This method is guaranteed to return a set of geocells having the same
    # resolution.
    # 
    # Args:
    #   bbox: A geotypes.Box indicating the bounding box being searched.
    #   cost_function: A function that accepts two arguments:
    #       * num_cells: the number of cells to search
    #       * resolution: the resolution of each cell to search
    #       and returns the 'cost' of querying against this number of cells
    #       at the given resolution.
    # 
    # Returns:
    #   A list of geocell strings that contain the given box.
    #
    def self.best_bbox_search_cells(bbox, cost_function)

      cell_ne = compute(bbox.north_east, MAX_GEOCELL_RESOLUTION)
      cell_sw = compute(bbox.south_west, MAX_GEOCELL_RESOLUTION)

      # The current lowest BBOX-search cost found; start with practical infinity.
      min_cost = 1e10000

      # The set of cells having the lowest calculated BBOX-search cost.
      min_cost_cell_set = nil

      # First find the common prefix, if there is one.. this will be the base
      # resolution.. i.e. we don't have to look at any higher resolution cells.
      min_resolution = common_prefix([cell_sw, cell_ne]).size

      # Iteravely calculate all possible sets of cells that wholely contain
      # the requested bounding box.
      (min_resolution..(MAX_GEOCELL_RESOLUTION + 1)).each do |cur_resolution|
        cur_ne = cell_ne[0...cur_resolution]
        cur_sw = cell_sw[0...cur_resolution]

        num_cells = interpolation_count(cur_ne, cur_sw)
        next if num_cells > MAX_FEASIBLE_BBOX_SEARCH_CELLS
          
        cell_set = interpolate(cur_ne, cur_sw).sort
        simplified_cells = []

        cost = cost_function.call(cell_set.size, cur_resolution)

        # TODO(romannurik): See if this resolution is even possible, as in the
        # future cells at certain resolutions may not be stored.
        if cost <= min_cost
          min_cost = cost
          min_cost_cell_set = cell_set
        else
          # Once the cost starts rising, we won't be able to do better, so abort.
          break
        end
      end

      min_cost_cell_set
    end

    # Determines whether the given cells are collinear along a dimension.
    # 
    # Returns True if the given cells are in the same row (column_test=False)
    # or in the same column (column_test=True).
    # 
    # Args:
    #   cell1: The first geocell string.
    #   cell2: The second geocell string.
    #   column_test: A boolean, where False invokes a row collinearity test
    #       and 1 invokes a column collinearity test.
    # 
    # Returns:
    #   A bool indicating whether or not the given cells are collinear in the given
    #   dimension.
    #
    def self.collinear(cell1, cell2, column_test)
      upto = [cell1.size, cell2.size].min - 1
      
      (0..upto).each do |i|
        x1, y1 = subdiv_xy(cell1[i])
        x2, y2 = subdiv_xy(cell2[i])
     
        # Check row collinearity (assure y's are always the same).
        return false if (!column_test && y1 != y2)

        # Check column collinearity (assure x's are always the same).
        return false if (column_test && x1 != x2)
      end
      
      true
    end

    # Calculates the grid of cells formed between the two given cells.
    # 
    # Generates the set of cells in the grid created by interpolating from the
    # given Northeast geocell to the given Southwest geocell.
    # 
    # Assumes the Northeast geocell is actually Northeast of Southwest geocell.
    # 
    # Arguments:
    #   cell_ne: The Northeast geocell string.
    #   cell_sw: The Southwest geocell string.
    # 
    # Returns:
    #   A list of geocell strings in the interpolation.
    # 
    def self.interpolate(cell_ne, cell_sw)
      # 2D array, will later be flattened.
      cell_set = [[cell_sw]]

      # First get adjacent geocells across until Southeast--collinearity with
      # Northeast in vertical direction (0) means we're at Southeast.
      while !collinear(cell_set.first.last, cell_ne, true)
        cell_tmp = adjacent(cell_set.first.last, [1, 0])
        cell_set.first << cell_tmp unless cell_tmp.nil?
      end

      # Then get adjacent geocells upwards.
      while cell_set.last.last != cell_ne
        cell_tmp_row = cell_set.last.map { |g| adjacent(g, [0, 1]) }
        cell_set << cell_tmp_row unless cell_tmp_row.first.nil? 
      end
      
      # Flatten cell_set, since it's currently a 2D array.
      cell_set.flatten
    end   

    # Computes the number of cells in the grid formed between two given cells.
    # 
    # Computes the number of cells in the grid created by interpolating from the
    # given Northeast geocell to the given Southwest geocell. Assumes the Northeast
    # geocell is actually Northeast of Southwest geocell.
    # 
    # Arguments:
    #   cell_ne: The Northeast geocell string.
    #   cell_sw: The Southwest geocell string.
    # 
    # Returns:
    #   An int, indicating the number of geocells in the interpolation.
    # 
    def self.interpolation_count(cell_ne, cell_sw)

      bbox_ne = compute_box(cell_ne)
      bbox_sw = compute_box(cell_sw)

      cell_lat_span = bbox_sw.north - bbox_sw.south
      cell_lon_span = bbox_sw.east - bbox_sw.west

      num_cols = ((bbox_ne.east - bbox_sw.west) / cell_lon_span).to_i
      num_rows = ((bbox_ne.north - bbox_sw.south) / cell_lat_span).to_i

      num_cols * num_rows
    end

    # Calculates all of the given geocell's adjacent geocells.
    # 
    # Args:
    #   cell: The geocell string for which to calculate adjacent/neighboring cells.
    # 
    # Returns:
    #   A list of 8 geocell strings and/or None values indicating adjacent cells.
    #
    def self.all_adjacents(cell)
      [NORTHWEST, NORTH, NORTHEAST, EAST, SOUTHEAST, SOUTH, SOUTHWEST, WEST].map { |d| adjacent(cell, d)}
    end

    # Calculates the geocell adjacent to the given cell in the given direction.
    # 
    # Args:
    #   cell: The geocell string whose neighbor is being calculated.
    #   dir: An (x, y) tuple indicating direction, where x and y can be -1, 0, or 1.
    #       -1 corresponds to West for x and South for y, and
    #        1 corresponds to East for x and North for y.
    #       Available helper constants are NORTH, EAST, SOUTH, WEST,
    #       NORTHEAST, NORTHWEST, SOUTHEAST, and SOUTHWEST.
    # 
    # Returns:
    #   The geocell adjacent to the given cell in the given direction, or None if
    #   there is no such cell.
    # 
    def self.adjacent(cell, dir)
      return nil if cell.nil?

      dx = dir[0]
      dy = dir[1]

      cell_adj_arr = cell.split(//)  # Split the geocell string characters into a list.
      i = cell_adj_arr.size - 1

      while i >= 0 && (dx != 0 || dy != 0)
        x, y = subdiv_xy(cell_adj_arr[i])

        # Horizontal adjacency.
        if dx == -1  # Asking for left.
          if x == 0  # At left of parent cell.
            x = GEOCELL_GRID_SIZE - 1  # Becomes right edge of adjacent parent.
          else
            x -= 1  # Adjacent, same parent.
            dx = 0  # Done with x.
          end
        elsif dx == 1  # Asking for right.
          if x == GEOCELL_GRID_SIZE - 1  # At right of parent cell.
            x = 0  # Becomes left edge of adjacent parent.
          else
            x += 1  # Adjacent, same parent.
            dx = 0  # Done with x.
          end
        end

        # Vertical adjacency.
        if dy == 1  # Asking for above.
          if y == GEOCELL_GRID_SIZE - 1  # At top of parent cell.
            y = 0  # Becomes bottom edge of adjacent parent.
          else
            y += 1  # Adjacent, same parent.
            dy = 0  # Done with y.
          end
        elsif dy == -1  # Asking for below.
          if y == 0  # At bottom of parent cell.
            y = GEOCELL_GRID_SIZE - 1  # Becomes top edge of adjacent parent.
          else
            y -= 1  # Adjacent, same parent.
            dy = 0  # Done with y.
          end
        end

        cell_adj_arr[i] = subdiv_char([x,y])
        i -= 1
      end
      
      # If we're not done with y then it's trying to wrap vertically,
      # which is a failure.
      return nil if dy != 0

      # At this point, horizontal wrapping is done inherently.
      cell_adj_arr.join('')
    end

    # Returns whether or not the given cell contains the given point.
    def self.contains_point(cell, point)
      compute(point, cell.size) == cell
    end

    # Returns the shortest distance between a point and a geocell bounding box.
    # 
    # If the point is inside the cell, the shortest distance is always to a 'edge'
    # of the cell rectangle. If the point is outside the cell, the shortest distance
    # will be to either a 'edge' or 'corner' of the cell rectangle.
    # 
    # Returns:
    #   The shortest distance from the point to the geocell's rectangle, in meters.
    #
    def self.point_distance(cell, point)
      bbox = compute_box(cell)

      between_w_e = bbox.west <= point.lon && point.lon <= bbox.east
      between_n_s = bbox.south <= point.lat && point.lat <= bbox.north

      if between_w_e
        if between_n_s
          # Inside the geocell.
          
          return [Geomodel::Math.distance(point, Geomodel::Types::Point.new(bbox.south, point.lon)),
                  Geomodel::Math.distance(point, Geomodel::Types::Point.new(bbox.north, point.lon)),
                  Geomodel::Math.distance(point, Geomodel::Types::Point.new(point.lat, bbox.east)),
                  Geomodel::Math.distance(point, Geomodel::Types::Point.new(point.lat, bbox.west))].min
        else
          return [Geomodel::Math.distance(point, Geomodel::Types::Point.new(bbox.south, point.lon)),
                  Geomodel::Math.distance(point, Geomodel::Types::Point.new(bbox.north, point.lon))].min
        end
      else
        if between_n_s
          return [Geomodel::Math.distance(point, Geomodel::Types::Point.new(point.lat, bbox.east)),
                  Geomodel::Math.distance(point, Geomodel::Types::Point.new(point.lat, bbox.west))].min
        else
          # TODO(romannurik): optimize
          return [Geomodel::Math.distance(point, Geomodel::Types::Point.new(bbox.south, bbox.east)),
                  Geomodel::Math.distance(point, Geomodel::Types::Point.new(bbox.north, bbox.east)),
                  Geomodel::Math.distance(point, Geomodel::Types::Point.new(bbox.south, bbox.west)),
                  Geomodel::Math.distance(point, Geomodel::Types::Point.new(bbox.north, bbox.west))].min
        end
      end
    end

    # Computes the geocell containing the given point to the given resolution.
    # 
    # This is a simple 16-tree lookup to an arbitrary depth (resolution).
    # 
    # Args:
    #   point: The geotypes.Point to compute the cell for.
    #   resolution: An int indicating the resolution of the cell to compute.
    # 
    # Returns:
    #   The geocell string containing the given point, of length <resolution>.
    # 
    def self.compute(point, resolution = MAX_GEOCELL_RESOLUTION)
      north = 90.0
      south = -90.0
      east = 180.0
      west = -180.0

      cell = ''
      while cell.size < resolution
        subcell_lon_span = (east - west) / GEOCELL_GRID_SIZE
        subcell_lat_span = (north - south) / GEOCELL_GRID_SIZE

        x = [(GEOCELL_GRID_SIZE * (point.longitude - west) / (east - west)).to_i, 
             GEOCELL_GRID_SIZE - 1].min
        y = [(GEOCELL_GRID_SIZE * (point.latitude - south) / (north - south)).to_i,
             GEOCELL_GRID_SIZE - 1].min

        cell += subdiv_char([x,y])

        south += subcell_lat_span * y
        north = south + subcell_lat_span

        west += subcell_lon_span * x
        east = west + subcell_lon_span
      end

      cell
    end

    # Computes the rectangular boundaries (bounding box) of the given geocell.
    # 
    # Args:
    #   cell: The geocell string whose boundaries are to be computed.
    # 
    # Returns:
    #   A geotypes.Box corresponding to the rectangular boundaries of the geocell.
    # 
    def self.compute_box(cell)
      return nil if cell.nil?

      bbox = Geomodel::Types::Box.new(90.0, 180.0, -90.0, -180.0)
      
      cell_copy = cell.clone

      while cell_copy.size > 0
        subcell_lon_span = (bbox.east - bbox.west) / GEOCELL_GRID_SIZE
        subcell_lat_span = (bbox.north - bbox.south) / GEOCELL_GRID_SIZE

        x, y = subdiv_xy(cell_copy[0])

        bbox = Geomodel::Types::Box.new(bbox.south + subcell_lat_span * (y + 1),
                                        bbox.west  + subcell_lon_span * (x + 1),
                                        bbox.south + subcell_lat_span * y,
                                        bbox.west  + subcell_lon_span * x)

        cell_copy.slice!(0)
      end
      
      bbox
    end

    # Returns whether or not the given geocell string defines a valid geocell.
    def self.is_valid(cell)
      !cell.nil? && 
      cell.size > 0 && 
      cell.split(//).inject(true) { |val, c| val && GEOCELL_ALPHABET.include?(c) }
    end

    # Calculates the immediate children of the given geocell.
    # 
    # For example, the immediate children of 'a' are 'a0', 'a1', ..., 'af'.
    # 
    def self.children(cell)
      GEOCELL_ALPHABET.split(//).map { |chr| cell + chr }
    end

    # Returns the (x, y) of the geocell character in the 4x4 alphabet grid.
    # NOTE: This only works for grid size 4.
    def self.subdiv_xy(char)
      char = GEOCELL_ALPHABET.index(char)
      [(char & 4) >> 1 | (char & 1) >> 0, (char & 8) >> 2 | (char & 2) >> 1]
    end

    # Returns the geocell character in the 4x4 alphabet grid at pos. (x, y).
    # NOTE: This only works for grid size 4.
    def self.subdiv_char(pos)
      GEOCELL_ALPHABET[(pos[1] & 2) << 2 | (pos[0] & 2) << 1 | (pos[1] & 1) << 1 | (pos[0] & 1) << 0]
    end
    
    def self.common_prefix(list)
      /\A(.*).*(\n\1.*)*\Z/.match(list.join("\n"))[1]
    end
  end
end